messaging/event-relay/src/main/solidity/SignedEventStore.sol

```diff
@@ -0,0 +1,150 @@
+        );
+    }
+
+    /**
+     * Relay event from one or more relayers.
+     *
+     * Note that replay protection and time outs are handled by the function call layer.
+     */
+    function relayEvent(
+        uint256 _blockchainId,
+        address _cbcAddress,
+        bytes calldata _encodedEvent,
+        bytes calldata _signature
+    ) external {
+        // Step 1: Check the signatures of signers.
+        address[] memory signers;
+        bytes32[] memory sigRs;
+        bytes32[] memory sigSs;
+        uint8[] memory sigVs;
+
+        uint32 len = BytesUtil.bytesToUint32(_signature, 0);
+        {
+            require(
+                _signature.length == LEN_OF_LEN + len * LEN_OF_SIG,
+                "Signature incorrect length"
+            );
+
+            signers = new address[](len);
+            sigRs = new bytes32[](len);
+            sigSs = new bytes32[](len);
+            sigVs = new uint8[](len);
+
+            uint256 offset = LEN_OF_LEN;
+            for (uint256 i = 0; i < len; i++) {
+                signers[i] = BytesUtil.bytesToAddress2(_signature, offset);
+                offset += 20;
+                sigRs[i] = BytesUtil.bytesToBytes32(_signature, offset);
+                offset += 32;
+                sigSs[i] = BytesUtil.bytesToBytes32(_signature, offset);
+                offset += 32;
+                sigVs[i] = BytesUtil.bytesToUint8(_signature, offset);
+                offset += 1;
+            }
+        }
+        registrar.verify(
+            _blockchainId,
+            signers,
+            sigRs,
+            sigSs,
+            sigVs,
+            _encodedEvent
+        );
+
+        // Step 2: Record who signed the event.
+        bytes32 signedEventDigest = keccak256(_encodedEvent);
+
+        for (uint256 i = 0; i < len; i++) {
+            require(
+                signedEvents[signedEventDigest].whoVotedMap[signers[i]] ==
+                    false,
+                "Relayer has already signed event"
+            );
+            signedEvents[signedEventDigest].whoVotedMap[signers[i]] = true;
+            signedEvents[signedEventDigest].whoVoted.push(signers[i]);
+        }
+
+        // Action the event if enough relayers have signed.
+        if (signedEvents[signedEventDigest].actioned) {
+            // There is no more to do. The signers have been recorded. This could
+            // be important for determining who to pay.
+            return;
+        }
+        uint256 threshold = registrar.getSigningThreshold(_blockchainId);
+        if (signedEvents[signedEventDigest].whoVoted.length >= threshold) {
+            // NOTE: In the call below, the _signature parameter isn't used. Could be cheaper to pass bytes("")
+            functionCall.crossCallHandler(
+                _blockchainId,
+                _cbcAddress,
+                _encodedEvent,
+                _signature
+            );
+            signedEvents[signedEventDigest].actioned = true;
```

Comment:
not really sure about this but, given the `crossCallHandler()` will probably result in calling an external application contract and that we don't have access controls on the `relayEvent()` function, might it be prudent to move the state-effect on line 147 before the function call on 141 so that if some kind of reentrancy happens presumably the check on `actioned` in 133 would help?

Reply:
Good point. It would seem unlikely, but possible.